Zend/zend.c

```diff
@@ -665,6 +665,22 @@ static void function_copy_ctor(zval *zv) /* {{{ */
 		}
 		func->common.arg_info = new_arg_info + 1;
 	}
+	if (old_func->common.attributes) {
+		zend_attribute *old_attr;
+
+		func->common.attributes = NULL;
+
+		ZEND_HASH_PACKED_FOREACH_PTR(old_func->common.attributes, old_attr) {
+			uint32_t i;
+			zend_attribute *attr;
+
+			attr = zend_add_attribute(&func->common.attributes, old_attr->name, old_attr->argc, old_attr->flags, old_attr->offset, old_attr->lineno);
```

Comment:
Right (regarding your most recent failure), ZVAL_DUP() does a non-persistent copy. The attributes copy must not be persistent either.

```suggestion
			attr = zend_add_attribute(&func->common.attributes, old_attr->name, old_attr->argc, old_attr->flags & ~ZEND_ATTRIBUTE_PERSISTENT, old_attr->offset, old_attr->lineno);
```

That way the attributes are not persistent (it's anyway a thread-local copy).

Reply:
Removing the `ZEND_ATTRIBUTE_PERSISTENT` flag causes ext/zend_test/tests/attribute_arguments.phpt to fail with:

```
php: /home/timwolla/Schreibtisch/php-src/ext/reflection/php_reflection.c:1139: read_attributes: Assertion `(((__ht)->u.flags & (1<<2)) != 0)' failed.
==242954== 
==242954== Process terminating with default action of signal 6 (SIGABRT)
==242954==    at 0x4A4C15B: raise (raise.c:51)
==242954==    by 0x4A2B858: abort (abort.c:79)
==242954==    by 0x4A2B728: __assert_fail_base.cold (assert.c:92)
==242954==    by 0x4A3CF05: __assert_fail (assert.c:101)
==242954==    by 0x4FCF61: read_attributes (php_reflection.c:1139)
==242954==    by 0x4FD30B: reflect_attributes (php_reflection.c:1204)
==242954==    by 0x5057FE: zim_ReflectionParameter_getAttributes (php_reflection.c:2785)
==242954==    by 0x79A422: ZEND_DO_FCALL_SPEC_RETVAL_USED_HANDLER (zend_vm_execute.h:1885)
==242954==    by 0x8162A1: execute_ex (zend_vm_execute.h:55535)
==242954==    by 0x81BA97: zend_execute (zend_vm_execute.h:59858)
==242954==    by 0x7575F4: zend_execute_scripts (zend.c:1767)
==242954==    by 0x699294: php_execute_script (main.c:2535)
```

Locally it works with the version in commit dbb619d7734b8788e1e62f1baaaf8894893349a8.

Reply:
Don't apply this change though if you use ZVAL_DUP() :-D I believe this here is more correct than using ZVAL_DUP(), but both should work for everything currently supported.

Reply:
Okay, then let's leave it as is.